File: frontend/src/lib/graphShape.ts
/**
 * What the verdict strip's ownership-network column is allowed to say.
 *
 * The column is an invitation into FullCheck with numbers on it, and numbers
 * on an invitation are a claim. `graph_shape` (backend, `_graph_shape` in
 * `routers/lookup.py`) counts the statements *this* check mapped — deduplicated
 * by statementId, because several sources describe the same party — and never
 * the GLEIF subsidiary total or anything FullCheck would go on to discover.
 *
 * This module decides when there is enough to render at all. Two rules, both
 * of which exist because the alternative reads as a finding:
 *
 * - **One company is not a network.** Every lookup produces at least the
 *   subject's own entity statement, so a card saying "1 company" would appear
 *   on every report including the ones where nothing else was found — an
 *   invitation to explore a graph with a single node in it.
 * - **Depth is only spoken when it was measured.** `COMPLEX_OWNERSHIP_LAYERS`
 *   carries the longest chain the risk layer actually walked. Absent, the
 *   column says nothing about depth rather than saying "1 layer deep", which
 *   would state a flat graph the check never established.
 */

import type { GraphShape } from "./api";

export interface NetworkSummary {
  companies: number;
  people: number;
  relationships: number;
  /** "four layers deep", or null when no chain was measured. */
  depthPhrase: string | null;
}

/** Smallest graph worth calling a network: the subject plus something else. */
const MIN_COMPANIES = 2;

const DEPTH_WORD = [
  "",
  "one",
  "two",
  "three",
  "four",
  "five",
  "six",
  "seven",
  "eight",
  "nine",
  "ten",
];

/** "four layers deep" — words up to ten, digits past it, so a 14-layer chain
 *  reads as "14 layers deep" rather than as an invented word. */
export function depthPhrase(depth: number | null | undefined): string | null {
  if (typeof depth !== "number" || !Number.isFinite(depth) || depth < 1) return null;
  const n = Math.floor(depth);
  const word = DEPTH_WORD[n] ?? String(n);
  return `${word} ${n === 1 ? "layer" : "layers"} deep`;
}

/**
 * The column's content, or `null` when there is nothing honest to put in it.
 *
 * `null` covers three cases that are all the same case: the event has not
 * landed, the backend is older than `graph_shape`, or the graph is just the
 * subject. In each of them the check has not established that a network
 * exists, so it must not offer to show one.
 */
export function networkSummary(shape?: GraphShape | null): NetworkSummary | null {
  if (!shape) return null;
  const companies = numberOr(shape.companies);
  const people = numberOr(shape.people);
  const relationships = numberOr(shape.relationships);
  // Relationships matter as well as companies: two unconnected entity
  // statements from two sources are two records of the same subject, not a
  // network, and the ownership view would render them side by side with no
  // edge between them.
  if (companies < MIN_COMPANIES && people === 0) return null;
  if (relationships === 0) return null;
  return { companies, people, relationships, depthPhrase: depthPhrase(shape.depth) };
}

function numberOr(value: unknown): number {
  return typeof value === "number" && Number.isFinite(value) && value > 0
    ? Math.floor(value)
    : 0;
}

File: backend/opencheck/graph_shape.py
"""How big a lookup's ownership-and-control graph is — the ``graph_shape`` event.

Split out of ``routers/lookup.py`` in Phase 246, unchanged: pure functions over
the merged BODS bundle, with no fetch and no clock. ``_count_parties``
collapses statements that share a published identifier, so one company
described by three sources counts once; ``_graph_shape`` adds the relationship
count and the longest chain the risk layer measured. ``routers/lookup.py``
re-exports all three names.
"""

from __future__ import annotations

from typing import Any


def _identifier_keys(statement: dict[str, Any]) -> list[str]:
    """The published identifiers on a statement, normalised for comparison."""
    details = statement.get("recordDetails") or {}
    keys: list[str] = []
    for ident in details.get("identifiers") or []:
        if not isinstance(ident, dict):
            continue
        scheme = ident.get("scheme") or ident.get("schemeName") or ""
        value = ident.get("id")
        if isinstance(value, str) and value.strip():
            keys.append(f"{scheme}:{value.strip()}".lower())
    return keys


def _count_parties(statements: list[dict[str, Any]]) -> int:
    """How many distinct parties a list of same-kind statements describes.

    Every mapper derives its ids as ``_stable_id(source_id, kind, local_id)``,
    so GLEIF's copy of a company and Companies House's copy have **different**
    ``statementId``s by construction. Counting statements and calling the total
    "companies" therefore overstated every graph where two sources describe the
    subject — which is nearly all of them.

    Records are joined into one party when they share a published identifier
    (``recordDetails.identifiers[]``), which is the evidence ``reconcile.py``
    already requires before it will assert cross-source corroboration. It is a
    transitive join, not a "pick one key" rule: GLEIF may publish only the LEI
    while Companies House publishes a company number *and* the LEI, so the two
    records agree on one identifier out of three and must still be one party.

    Names are deliberately **not** used. A name match is not an identity claim
    anywhere else in OpenCheck — ``possibly_same_entities`` exists precisely to
    hold name-only pairs out of the graph and hand them to a human — and it
    must not become one here just because it would make a number smaller.

    The consequence is that the figure is an **upper bound**: two records of
    one company sharing no identifier stay two. That is the safe direction for
    an invitation into FullCheck, whose job is to go and resolve exactly those.
    """
    parent: dict[str, str] = {}

    def find(x: str) -> str:
        parent.setdefault(x, x)
        while parent[x] != x:
            parent[x] = parent[parent[x]]
            x = parent[x]
        return x

    def union(a: str, b: str) -> None:
        ra, rb = find(a), find(b)
        if ra != rb:
            parent[ra] = rb

    roots: list[str] = []
    for index, statement in enumerate(statements):
        sid = statement.get("statementId")
        node = f"stmt:{sid}" if isinstance(sid, str) else f"stmt:#{index}"
        find(node)
        roots.append(node)
        for key in _identifier_keys(statement):
            union(node, f"id:{key}")

    return len({find(node) for node in roots})


def _graph_shape(
    bods: list[dict[str, Any]], signals: list[dict[str, Any]]
) -> dict[str, int | None]:
    """How big the ownership-and-control graph on this page actually is.

    The report's third verdict column invites the reader into FullCheck, and an
    invitation with no numbers on it is a button. These are the numbers the
    check has *already earned*: the parties in the merged BODS bundle — the
    same bundle ``/export`` ships and the risk engine assessed — collapsed by
    ``_count_parties`` so one company described by three sources counts once.

    It deliberately does **not** reach for the GLEIF subsidiary total or
    anything FullCheck would go on to discover. Those are a different scope,
    and a sentence that mixes "what we have" with "what we might find" is the
    same overclaim as a progress bar that runs ahead of its stream.

    ``depth`` is the longest ownership chain the risk layer actually measured
    (``COMPLEX_OWNERSHIP_LAYERS`` carries it as ``evidence.longest_path``), or
    ``None`` when the signal did not fire — never a guess, and never 0, which
    would render as a flat graph.
    """
    entities = [s for s in bods if s.get("recordType") == "entity"]
    persons = [s for s in bods if s.get("recordType") == "person"]
    relationships = {
        s.get("statementId")
        for s in bods
        if s.get("recordType") == "relationship" and isinstance(s.get("statementId"), str)
    }

    depth: int | None = None
    for signal in signals:
        if signal.get("code") != "COMPLEX_OWNERSHIP_LAYERS":
            continue
        path = (signal.get("evidence") or {}).get("longest_path")
        if isinstance(path, list) and path:
            depth = max(depth or 0, len(path))
    return {
        "companies": _count_parties(entities),
        "people": _count_parties(persons),
        "relationships": len(relationships),
        "depth": depth,
    }
